### ISSUE TITLE
[Users of NetworkPeer + NetworkPeer] Beyond NetworkPeer refactoring - separate Dispose from Disconnect to avoid cyclic calls

### ISSUE THREAD

**Issue by Aprogiena** [stratisproject/StratisBitcoinFullNode#1090]
Currently various behaviors call disconnect/dispose on network peer and this leads to cyclic calls like 

`state changed -> on state changed callback -> dispose peer -> disconnect peer ->state changed to offline -> on state changed callback -> dispose peer`

to prevent those, we need to do following steps:
- separate disconnection from disposal (disconnection would only mean doing what is now in `NetworkPeerConnection.Disconnect`)
- allow disconnection from callbacks but not disposals (this is optional, but a good think to have to see if we fixed those everywhere)
- each component that creates network peer has to manage the list of peers and dispose properly 

kind of a big task, which won't be included in NetworkPeer refactoring series

once this is done, we can relax (remove functionality) AsyncQueue and maybe even AsyncExecutionEvent, which are now coded in a way that this cyclic calling is allowed

then we can very much refactor NetworkPeer's `Disconnect` and NetworkPeerConnection's `Shutdown`, and `Dispose`, those could be much much simpler after this is no longer needed

**Comment by Aprogiena** [comment]
we would maybe want then to introduce OnDisconnect event in NetworkPeer so that those components owning the peer can get notified when this happens (i.e. completely different behavior calls disconnect on their peer) and so they can initiate disposal, but either they would manage themselves that or we would call this OnDisconnect in a non-synchronized fashion, so that we allow disposals from those

**Comment by noescape00** [comment]
I think we can avoid cyclic calls without removing the ability to call Dispose from the callback. 

So what we can do is that: 
1) Save current `NetworkPeerState` when we start callback execution. 
2) After each callback is executed check current state against saved state. If they are not equal- stop current execution because newer state will run callbacks again. 

What's the point of notifying peers about a new state when this state is already outdated?

**Comment by Aprogiena** [comment]
I don't think that would be safe to do, I can imagine scenarios where component relies on something being that when the peer is switched to certain state and if you just about the execution it won't happen.

So to me it seems like such a counterintuitive feature of the callback could lead to more problems than it would solve.

And also because the behaviours are not very well designed, I would rather see refactoring that would clean it up

**Comment by noescape00** [comment]
>I don't think that would be safe to do, I can imagine scenarios where component relies on something being that when the peer is switched to certain state and if you just about the execution it won't happen.

I don't think we have behaviors that rely on a certain sequence of states being signaled in order to work correctly. 
Though if we want to ensure that we can't jump form one state to a random other we probably should guard against that. 

>And also because the behaviours are not very well designed, I would rather see refactoring that would clean it up

Ok, I don't mind going with the "big redesign" approach. 

I'll think on the best way of approaching that and post new design suggestion here.

**Comment by Aprogiena** [comment]
I think that is the problem, no one knows. And possibly even bigger problem is that such hidden feature could turn against you in the future.

The refactoring here would not be that much extreme, the number of behaviours we need to attach is reasonable. We only need extreme amount of testing one this is done and we should be fine.

**Comment by Aprogiena** [comment]
I kind of disagree the full step one, this step is unnecessary to fix the issue, it is like completely different thing that is unrelated. Moreover it is already implemented in different class, see network peer listener

**Comment by Aprogiena** [comment]
> Instead of 'Dispose(string reason, Exception exception = null)', 'Dispose()' and 'Disconnect(string reason, Exception exception = null)' use single "DisconnectAndDispose".
The reason for that is: we can't disconnect without disposing. After peer is disconnected it's no longer usable so we should do it in a single method.

I also disagree with this one, we should disconnect separately because it is exactly this that is causing the mess at  this moment


As for other suggestions, I will need to look into the code to be able to comment on those

**Comment by noescape00** [comment]
>I also disagree with this one, we should disconnect separately because it is exactly this that is causing the mess at this moment

With those change in place (less states and single DisconnectAndDispose) there won't be place for mess that is currently going on. 

When peer is being disposed we will first wait until all behaviors are disposed and after that we start our own disposure.

I'd agree with you if we would be able to do Disconnect and then Connect again. Right now Disconnect is 'dispose, but not completely'

**Comment by noescape00** [comment]
>Moreover it is already implemented in different class, see network peer listener

peer listener implements **pull** mechanic so in order to use it we would need to implement a loop that would continuously ask for a specific payload. And if we need 2 different payloads we need 2 loops or implement WhenAny logic. 

What I want to do is to implement a **push** mechanic which will make payload processing extremely easy on the user side: create a private method that accepts the payload and register this method on the initialization- that's it. 

>I kind of disagree the full step one, this step is unnecessary to fix the issue, it is like completely different thing that is unrelated. 

It is unrelated to the cyclic calls problem but it is related to the refactoring of the behaviors. As you said: 

>And also because the behaviours are not very well designed, I would rather see refactoring that would clean it up

**Comment by Aprogiena** [comment]
I would like to see behaviours refactored, but not as a part of this issue. Let's focus here on removing the cycles and let's keep it minimal.

I will explain on Skype why I don't think that the suggested mechanism of registration is good here
